File: packages/east-py-datascience/src/east_py_datascience/simanneal/simanneal.py
"""Simulated Annealing platform functions for East.

Provides discrete optimization using the simanneal library.
Ideal for combinatorial problems like TSP, scheduling, and subset selection.
"""

from typing import Any, Callable
import random

from east.runtime.platform import PlatformFunction
from east.types.types import (
    ArrayType,
    BooleanType,
    FloatType,
    FunctionType,
    IntegerType,
    OptionType,
    StructType,
    VariantType,
)
from east.types.values import EastArray, EastStruct, EastVariant, is_east_variant

# ============================================================================
# Type Definitions
# ============================================================================

# Discrete state type (int_array or bool_array)
DiscreteStateType = VariantType(
    [
        ("int_array", ArrayType(IntegerType)),
        ("bool_array", ArrayType(BooleanType)),
    ]
)

# Energy function type: state -> score
EnergyFunctionType = FunctionType([DiscreteStateType], FloatType)

# Move function type: state -> neighbor
MoveFunctionType = FunctionType([DiscreteStateType], DiscreteStateType)

# Permutation energy function type
PermutationEnergyType = FunctionType([ArrayType(IntegerType)], FloatType)

# Subset energy function type
SubsetEnergyType = FunctionType([ArrayType(BooleanType)], FloatType)

# Annealing configuration
AnnealConfigType = StructType(
    [
        ("t_max", OptionType(FloatType)),
        ("t_min", OptionType(FloatType)),
        ("steps", OptionType(IntegerType)),
        ("updates", OptionType(IntegerType)),
        ("auto_schedule", OptionType(FloatType)),
        ("random_state", OptionType(IntegerType)),
    ]
)

# Annealing result
AnnealResultType = StructType(
    [
        ("best_state", DiscreteStateType),
        ("best_energy", FloatType),
        ("steps_taken", IntegerType),
        ("success", BooleanType),
    ]
)


# ============================================================================
# Helper Functions
# ============================================================================


def _get_option(opt: EastVariant | None, default: Any) -> Any:
    """Extract value from Option variant, returning default if None."""
    if opt is None:
        return default
    if is_east_variant(opt) and opt.type == "some":
        return opt.value
    return default


# ============================================================================
# Platform Function Implementations
# ============================================================================


def simanneal_optimize_impl(
    initial_state: EastVariant,
    energy_fn: Callable[[EastVariant], float],
    move_fn: Callable[[EastVariant], EastVariant],
    config: EastStruct,
) -> EastStruct:
    """Run simulated annealing with custom energy and move functions."""
    from simanneal import Annealer

    # Set random seed if provided
    random_state = _get_option(config.get("random_state"), None)
    if random_state is not None:
        random.seed(int(random_state))

    class EastAnnealer(Annealer):
        """Annealer that wraps East energy and move functions."""

        copy_strategy = "method"

        def __init__(self, state, energy_fn, move_fn):
            self.energy_fn = energy_fn
            self.move_fn = move_fn
            super().__init__(state)

        def copy_state(self, state):
            """Custom copy for EastVariant - states are immutable."""
            # EastVariant is immutable, so we can just return it
            return state

        def move(self):
            """Generate neighbor state using East move function."""
            self.state = self.move_fn(self.state)

        def energy(self):
            """Calculate energy using East energy function."""
            return self.energy_fn(self.state)

    # Create annealer instance
    annealer = EastAnnealer(initial_state, energy_fn, move_fn)

    # Configure schedule
    t_max = _get_option(config.get("t_max"), None)
    if t_max is not None:
        annealer.Tmax = float(t_max)

    t_min = _get_option(config.get("t_min"), None)
    if t_min is not None:
        annealer.Tmin = float(t_min)

    steps = _get_option(config.get("steps"), None)
    if steps is not None:
        annealer.steps = int(steps)

    updates = _get_option(config.get("updates"), None)
    if updates is not None:
        annealer.updates = int(updates)
    else:
        annealer.updates = 0  # Suppress output by default

    # Auto-calibrate if requested
    auto_minutes = _get_option(config.get("auto_schedule"), None)
    if auto_minutes is not None:
        schedule = annealer.auto(minutes=float(auto_minutes))
        annealer.set_schedule(schedule)

    # Run optimization
    best_state, best_energy = annealer.anneal()

    return EastStruct(
        {
            "best_state": best_state,
            "best_energy": float(best_energy),
            "steps_taken": int(annealer.steps),
            "success": True,
        }
    )


def simanneal_optimize_permutation_impl(
    initial_perm: EastArray,
    energy_fn: Callable[[EastArray], float],
    config: EastStruct,
) -> EastStruct:
    """Run simulated annealing on a permutation using swap moves."""
    from simanneal import Annealer

    # Set random seed if provided
    random_state = _get_option(config.get("random_state"), None)
    if random_state is not None:
        random.seed(int(random_state))

    # Convert to list for efficient swapping
    state_list = [int(x) for x in initial_perm]
    n = len(state_list)

    class PermutationAnnealer(Annealer):
        """Annealer for permutation problems with swap moves."""

        copy_strategy = "slice"  # Efficient list copying

        def __init__(self, state, energy_fn, n):
            self.energy_fn = energy_fn
            self.n = n
            super().__init__(state)

        def move(self):
            """Swap two random elements."""
            i = random.randint(0, self.n - 1)
            j = random.randint(0, self.n - 1)
            self.state[i], self.state[j] = self.state[j], self.state[i]

        def energy(self):
            """Calculate energy from permutation."""
            perm_array: EastArray = EastArray(IntegerType, self.state)
            return self.energy_fn(perm_array)

    annealer = PermutationAnnealer(state_list, energy_fn, n)

    # Configure schedule
    t_max = _get_option(config.get("t_max"), None)
    if t_max is not None:
        annealer.Tmax = float(t_max)

    t_min = _get_option(config.get("t_min"), None)
    if t_min is not None:
        annealer.Tmin = float(t_min)

    steps = _get_option(config.get("steps"), None)
    if steps is not None:
        annealer.steps = int(steps)

    updates = _get_option(config.get("updates"), None)
    if updates is not None:
        annealer.updates = int(updates)
    else:
        annealer.updates = 0

    auto_minutes = _get_option(config.get("auto_schedule"), None)
    if auto_minutes is not None:
        schedule = annealer.auto(minutes=float(auto_minutes))
        annealer.set_schedule(schedule)

    # Run optimization
    best_state_list, best_energy = annealer.anneal()

    return EastStruct(
        {
            "best_state": EastVariant(
                "int_array", EastArray(IntegerType, best_state_list)
            ),
            "best_energy": float(best_energy),
            "steps_taken": int(annealer.steps),
            "success": True,
        }
    )


def simanneal_optimize_subset_impl(
    initial_selection: EastArray,
    energy_fn: Callable[[EastArray], float],
    config: EastStruct,
) -> EastStruct:
    """Run simulated annealing on a subset selection using bit-flip moves."""
    from simanneal import Annealer

    # Set random seed if provided
    random_state = _get_option(config.get("random_state"), None)
    if random_state is not None:
        random.seed(int(random_state))

    state_list = [bool(x) for x in initial_selection]
    n = len(state_list)

    class SubsetAnnealer(Annealer):
        """Annealer for subset selection with bit-flip moves."""

        copy_strategy = "slice"

        def __init__(self, state, energy_fn, n):
            self.energy_fn = energy_fn
            self.n = n
            super().__init__(state)

        def move(self):
            """Flip a random bit."""
            i = random.randint(0, self.n - 1)
            self.state[i] = not self.state[i]

        def energy(self):
            """Calculate energy from selection."""
            selection_array: EastArray = EastArray(BooleanType, self.state)
            return self.energy_fn(selection_array)

    annealer = SubsetAnnealer(state_list, energy_fn, n)

    # Configure schedule
    t_max = _get_option(config.get("t_max"), None)
    if t_max is not None:
        annealer.Tmax = float(t_max)

    t_min = _get_option(config.get("t_min"), None)
    if t_min is not None:
        annealer.Tmin = float(t_min)

    steps = _get_option(config.get("steps"), None)
    if steps is not None:
        annealer.steps = int(steps)

    updates = _get_option(config.get("updates"), None)
    if updates is not None:
        annealer.updates = int(updates)
    else:
        annealer.updates = 0

    auto_minutes = _get_option(config.get("auto_schedule"), None)
    if auto_minutes is not None:
        schedule = annealer.auto(minutes=float(auto_minutes))
        annealer.set_schedule(schedule)

    best_state_list, best_energy = annealer.anneal()

    return EastStruct(
        {
            "best_state": EastVariant(
                "bool_array", EastArray(BooleanType, best_state_list)
            ),
            "best_energy": float(best_energy),
            "steps_taken": int(annealer.steps),
            "success": True,
        }
    )


# ============================================================================
# Platform Function Registration
# ============================================================================

simanneal_impl = [
    PlatformFunction(
        name="simanneal_optimize",
        inputs=[
            DiscreteStateType,
            EnergyFunctionType,
            MoveFunctionType,
            AnnealConfigType,
        ],
        output=AnnealResultType,
        type="sync",
        fn=simanneal_optimize_impl,
    ),
    PlatformFunction(
        name="simanneal_optimize_permutation",
        inputs=[
            ArrayType(IntegerType),
            PermutationEnergyType,
            AnnealConfigType,
        ],
        output=AnnealResultType,
        type="sync",
        fn=simanneal_optimize_permutation_impl,
    ),
    PlatformFunction(
        name="simanneal_optimize_subset",
        inputs=[
            ArrayType(BooleanType),
            SubsetEnergyType,
            AnnealConfigType,
        ],
        output=AnnealResultType,
        type="sync",
        fn=simanneal_optimize_subset_impl,
    ),
]


__all__ = [
    "simanneal_impl",
    "DiscreteStateType",
    "EnergyFunctionType",
    "MoveFunctionType",
    "PermutationEnergyType",
    "SubsetEnergyType",
    "AnnealConfigType",
    "AnnealResultType",
]

File: packages/east-py-datascience/src/simanneal/simanneal.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/**
 * Discrete optimization using Simulated Annealing.
 *
 * Provides combinatorial optimization for discrete state spaces using the
 * simanneal library. Ideal for:
 * - Permutation problems (TSP, scheduling)
 * - Subset selection (feature selection, knapsack)
 * - Assignment problems
 * - Any discrete optimization where gradient methods don't apply
 *
 * @packageDocumentation
 */

import {
    East,
    StructType,
    VariantType,
    OptionType,
    IntegerType,
    FloatType,
    BooleanType,
    FunctionType,
    VectorType,
} from "@elaraai/east";

// ===========================================
// State Types
// ===========================================

/**
 * Discrete state type for simulated annealing.
 *
 * Supports different representations for combinatorial problems:
 * - int_array: Permutations, assignments, sequences of integers
 * - bool_array: Subset selections, binary decisions
 */
export const DiscreteStateType = VariantType({
    /** Integer array state (permutations, assignments) */
    int_array: VectorType(IntegerType),
    /** Boolean array state (subset selection) */
    bool_array: VectorType(BooleanType),
});

// ===========================================
// Function Types
// ===========================================

/**
 * Energy function type: state -> score.
 *
 * Computes the objective value (energy) for a given state.
 * Lower energy is better (minimization).
 */
export const EnergyFunctionType = FunctionType(
    [DiscreteStateType],
    FloatType
);

/**
 * Move function type: state -> neighbor state.
 *
 * Generates a random neighbor of the current state.
 * Should make small, local changes (e.g., swap two elements).
 */
export const MoveFunctionType = FunctionType(
    [DiscreteStateType],
    DiscreteStateType
);

/**
 * Permutation energy function type.
 *
 * Specialized for permutation-based optimization.
 */
export const PermutationEnergyType = FunctionType(
    [VectorType(IntegerType)],
    FloatType
);

/**
 * Subset energy function type.
 *
 * Specialized for subset selection optimization.
 */
export const SubsetEnergyType = FunctionType(
    [VectorType(BooleanType)],
    FloatType
);

// ===========================================
// Configuration Types
// ===========================================

/**
 * Simulated annealing configuration.
 */
export const AnnealConfigType = StructType({
    /** Starting temperature (default: 25000.0) */
    t_max: OptionType(FloatType),
    /** Ending temperature (default: 2.5) */
    t_min: OptionType(FloatType),
    /** Total iterations (default: 50000) */
    steps: OptionType(IntegerType),
    /** Progress report frequency (default: 0 = silent) */
    updates: OptionType(IntegerType),
    /** Minutes for auto-calibration (default: none) */
    auto_schedule: OptionType(FloatType),
    /** Random seed for reproducibility */
    random_state: OptionType(IntegerType),
});

// ===========================================
// Result Types
// ===========================================

/**
 * Simulated annealing result.
 */
export const AnnealResultType = StructType({
    /** Best state found */
    best_state: DiscreteStateType,
    /** Energy of best state */
    best_energy: FloatType,
    /** Actual iterations performed */
    steps_taken: IntegerType,
    /** Whether optimization completed successfully */
    success: BooleanType,
});

// ===========================================
// Platform Functions
// ===========================================

/**
 * Run simulated annealing on a discrete state space.
 *
 * Uses custom energy and move functions for flexible optimization.
 *
 * @param initial_state - Starting state
 * @param energy_fn - Function to compute state energy (lower is better)
 * @param move_fn - Function to generate neighbor states
 * @param config - Annealing schedule configuration
 * @returns Result with best state and energy
 */
export const simanneal_optimize = East.platform(
    "simanneal_optimize",
    [
        DiscreteStateType,
        EnergyFunctionType,
        MoveFunctionType,
        AnnealConfigType,
    ],
    AnnealResultType
);

/**
 * Run simulated annealing on a permutation with swap moves.
 *
 * Convenience function for permutation-based problems (TSP, scheduling).
 * Automatically uses swap moves to generate neighbors.
 *
 * @param initial_perm - Starting permutation
 * @param energy_fn - Function to compute permutation energy
 * @param config - Annealing schedule configuration
 * @returns Result with best permutation and energy
 */
export const simanneal_optimize_permutation = East.platform(
    "simanneal_optimize_permutation",
    [
        VectorType(IntegerType),
        PermutationEnergyType,
        AnnealConfigType,
    ],
    AnnealResultType
);

/**
 * Run simulated annealing on a subset selection with bit-flip moves.
 *
 * Convenience function for subset selection problems (feature selection, knapsack).
 * Automatically uses bit-flip moves to generate neighbors.
 *
 * @param initial_selection - Starting selection (boolean array)
 * @param energy_fn - Function to compute selection energy
 * @param config - Annealing schedule configuration
 * @returns Result with best selection and energy
 */
export const simanneal_optimize_subset = East.platform(
    "simanneal_optimize_subset",
    [
        VectorType(BooleanType),
        SubsetEnergyType,
        AnnealConfigType,
    ],
    AnnealResultType
);

// ===========================================
// Grouped Export
// ===========================================

/**
 * Type definitions for Simulated Annealing.
 */
export const SimAnnealTypes = {
    /** Discrete state type */
    DiscreteStateType,
    /** Energy function type */
    EnergyFunctionType,
    /** Move function type */
    MoveFunctionType,
    /** Permutation energy function type */
    PermutationEnergyType,
    /** Subset energy function type */
    SubsetEnergyType,
    /** Configuration type */
    ConfigType: AnnealConfigType,
    /** Result type */
    ResultType: AnnealResultType,
} as const;

/**
 * Discrete optimization using Simulated Annealing.
 *
 * Provides combinatorial optimization for discrete state spaces.
 * Simulated annealing is a probabilistic technique that can escape
 * local minima by occasionally accepting worse solutions.
 *
 * Ideal for:
 * - Permutation problems (TSP, job scheduling)
 * - Subset selection (feature selection, knapsack)
 * - Assignment problems
 * - Any discrete optimization with many local minima
 *
 * @example
 * ```ts
 * import { East, FloatType, variant } from "@elaraai/east";
 * import { SimAnneal } from "@elaraai/east-py-datascience";
 *
 * // TSP: minimize total route distance
 * const energy = East.function(
 *     [VectorType(IntegerType)],
 *     FloatType,
 *     ($, route) => {
 *         // Calculate total route distance
 *         return $.return(totalDistance);
 *     }
 * );
 *
 * const config = $.let({
 *     t_max: variant("some", 10000.0),
 *     t_min: variant("some", 1.0),
 *     steps: variant("some", 50000n),
 *     updates: variant("none", null),
 *     auto_schedule: variant("none", null),
 *     random_state: variant("some", 42n),
 * });
 *
 * const result = SimAnneal.optimizePermutation(initial, energy, config);
 * ```
 */
export const SimAnneal = {
    /**
     * Run simulated annealing with custom energy and move functions.
     */
    optimize: simanneal_optimize,

    /**
     * Run simulated annealing on a permutation with swap moves.
     */
    optimizePermutation: simanneal_optimize_permutation,

    /**
     * Run simulated annealing on a subset selection with bit-flip moves.
     */
    optimizeSubset: simanneal_optimize_subset,

    /**
     * Type definitions for SimAnneal functions.
     */
    Types: SimAnnealTypes,
} as const;
